### PR TITLE
ValueCodable protocol objects

### PR DESCRIPTION
**Pull request by jgallagher**
:warning: do not merge :warning: 

Not meant as a real solution (probably), just a demo of something that apparently works

### CODE REVIEW

File: ValueCodable.playground/Sources/ValueCodable.swift [hunk 1]
```diff
@@ -88,6 +88,22 @@ extension NSCoder {
         }
     }
 
+    public func decodeDynamicValue<Value, T0: ValueCodable, T1: ValueCodable>(ofType _: Value.Type, forKey key: String, _ t0: T0.Type, _ t1: T1.Type) -> Value? {
```
Comment (jgallagher) [inline ValueCodable.playground/Sources/ValueCodable.swift:91]: Without variadic generics, would need to write an overload of this that took the number of concrete types you have

File: ValueCodable.playground/Contents.swift [hunk 2]
```diff
@@ -127,17 +167,20 @@ extension Person: ValueCodable, Equatable {
         coder.encodeValue(favoriteColor, forKey: Keys.FavoriteColor.rawValue)
         coder.encodeValue(weightClass, forKey: Keys.WeightClass.rawValue)
         coder.encodeValue(zodiacSign, forKey: Keys.ZodiacSign.rawValue)
+        dummy.encodeDynamic(with: coder, forKey: "dummy")
```
Comment (mdmathias) [inline ValueCodable.playground/Contents.swift:170]: Is there a reason why you're not using `Keys.Dummy.rawValue`?

File: ValueCodable.playground/Contents.swift [hunk 3]
```diff
@@ -127,17 +167,20 @@ extension Person: ValueCodable, Equatable {
...
     }
 
     init?(coder aDecoder: NSCoder) {
         guard let fullName = aDecoder.decodeValue(ofType: String.self, forKey: Keys.FullName.rawValue),
             favoriteColor = aDecoder.decodeValue(ofType: ColorChoice.self, forKey: Keys.FavoriteColor.rawValue),
             weightClass = aDecoder.decodeValue(ofType: WeightClass.self, forKey: Keys.WeightClass.rawValue),
-            zodiacSign = aDecoder.decodeValue(ofType: Zodiac.self, forKey: Keys.ZodiacSign.rawValue) else { return nil }
+            zodiacSign = aDecoder.decodeValue(ofType: Zodiac.self, forKey: Keys.ZodiacSign.rawValue),
+            dummy = aDecoder.decodeDynamicValue(ofType: DummyProtocolObject.self, forKey: "dummy", Foo.self, Bar.self) else { return nil }
```
Comment (mdmathias) [inline ValueCodable.playground/Contents.swift:178]: Same question here.

File: ValueCodable.playground/Contents.swift (hunk 2 above)
Comment (jgallagher) [inline ValueCodable.playground/Contents.swift:170]: No. Also, it's not a real PR

File: ValueCodable.playground/Contents.swift [hunk 4]
```diff
@@ -108,18 +108,58 @@ enum Zodiac: UnicodeScalar, ValueCodable {
     case Pisces = "♓"
 }
 
+protocol DummyProtocolObject: ValueCodable {
+    var x: Int { get }
+}
+
+extension DummyProtocolObject {
+    func encodeDynamic(with coder: NSCoder, forKey key: String) {
+        coder.encodeDynamicValue(self, asType: self.dynamicType, forKey: key)
```
Comment (jgallagher) [inline ValueCodable.playground/Contents.swift:117]: This line is the most suspect of them all. We're calling a generic function with a type bound to something that isn't known until runtime. I'm sorta surprised this works at all. This is taking advantage of the whole "Swift doesn't always specialize generic functions" thing that's usually annoying.

**Comment by zwaldowski** [comment]
_uncontrollable shivering_

**Comment by zwaldowski** [comment]
I've updated with the `ValueCodable` currently shipping against Swift 4 in (mumble), any other effort investment in this kind of thing should go into `Codable` instead.

**Pull request closed without merging**